_posts/2017-04-30-React-Native-for-iOS-devs.markdown

```diff
@@ -0,0 +1,248 @@
+
+- JS side: Instant, can run at the same time as your app
+- Native side: Requires stopping your app, running tests, then restarting the work
+
+We had a few native tests, but very quickly we stopped running them. 
+
+- We use Jest for all these reasons (quote JS post)
+
+- CI process is just a linux box
+- CI takes ~3m
+
+
+Deployment
+
+- Because JS is separate from app, JS can be updated separately
+- JS is not dynamically changing application via swizzling etc - just new JS talking to existing native code
+- This means you _can_ ship a different version of the JS to your app, but not all features can be shipped to old clients
+- We don't do it, for our ~month cadence, 2 day review time is OK
+- We do use it for having betas using different builds of the JS runtime. No need to deploy to testflight on every commit when we can ship just the JS and make our own commit chooser.
+- Deployment is tricky because you have two version number for _your_ RN: the version of the components, and the version of your native bridge
+
+Doing it right per platform
+
+- RN gives you the ability to think in cross-platform
+- Most devs are JS people trying to ship to Android/iOS
+- They're interested in getting it done vs getting it right
+- show my navigation issue
+
+- Doing it right will inevitably require native code
+- RN is a focused UI framework 
+- Does not hinder you from making apps with traditionally native features: e.g. NSUserActivity, Spotlight etc
+- Making some of these features requires crossing the bridge back to native land and having the work done there
+- _We_ still think in terms of UIViewControllers which have a react view tree, not there is an app react component with sub-react components as view controllers.
+
+- RN provides some pretty clever ways to handle cross-platform 
+
+- note somewhere that our imageviews use SDWebImage to share cache
+
+Create React App
+
+Ironically after this, one of the biggest projects to happen in RN in the last 6 months is CRNA. This is the super easy to get started RN experience. It's what we used earlier to get started.
+
+- Makes one big assumption: you will not write native code.
+- Remove this assumption and you can really tighten the tooling at JS level.
+- Leaves the native side of it to Expo
+- Expo is another company, but all the code is OSS - and they seem nice
+- Can handle compiling and shipping your app for you from the cloud
+- You can eject from CRNA to your own app at any time
+
+Animations
+
+- React Native probably isn't the right tooling for building something like Garageband. 
+- Excels at handling user state -> user interface
+- However, most apps are using pretty small animations here and there and React Native handles those _really, really well_.
+- Provides a tonne of primitives to allow writing declaritive animations
+
+Places where React Native hasn't fit for us
+
+- So far, nowhere. Lols. Can't be true overall though.
+- Something like our Kiosk app isn't an amazing fit
```

Comment:
Writing this started the same thought process TBH. 

We already "bridge" the Card Reader into an RX-like API, doing that to an event emitter in JS isn't much further of an abstraction.